File: programmers/lv3/미로 탈출 명령어.py
def solution(n, m, x, y, r, c, k):
    answer = ''
    
    dy, dx = r - x, c - y
    cy, cx = x, y
    
    while k > 0:
        if k == 1:
            if abs(dx + dy) != 1: return 'impossible'
            if dx == 1: answer += 'r'
            elif dx == -1: answer += 'l'
            elif dy == 1: answer += 'd'
            else: answer += 'u'
        # dy가 0보다 클 때는 무조건 아래로 이동
        elif dy > 0:
            answer += 'd'
            cy += 1
            dy -= 1
        # dy가 0보다 작거나 같음, 장난칠 여유 있음
        elif dy <= 0 and cy < n and k >= abs(dx) + (abs(dy) + 2):
            answer += 'd'
            cy += 1
            dy -= 1
        # dx가 0보다 작으면 무조건 왼쪽으로 이동
        elif dx < 0:
            answer += 'l'
            cx -= 1
            dx += 1
        # dx가 0 크거나 같음, 장난칠 여유 있음
        elif dx >= 0 and k >= (abs(dx) + 2) + abs(dy):
            # 왼쪽으로 더 갈 수 있음
            if cx > 1:
                answer += 'l'
                cx -= 1
                dx += 1
            # 왼쪽으로 더 갈 수 없음
            else:
                answer += 'r'
                cx += 1
                dx -= 1
        # dx가 0보다 큼, 장난칠 여유 없음
        elif dx > 0 and k < (abs(dx) + 2) + abs(dy):
            answer += 'r'
            cx += 1
            dx -= 1
        else:
            answer += 'u'
            cy -= 1
            dy += 1
        
        k -= 1
    
    return answer

File: programmers/lv3/미로 탈출 명령어.java
class Solution {
    public String solution(int R, int C, int r, int c, int tr, int tc, int k) {
        StringBuilder answer = new StringBuilder();
        r--;
        c--;
        tr--;
        tc--;
        
        while(k-- > 0) {
            // 아래로 내려갈 수 있으면 무조건 아래로
            if(r < tr || (r + 1 < R && Math.abs((r + 1) - tr) + Math.abs(c - tc) <= k)) {
                r++;
                answer.append("d");
                continue;
            }
            // 왼쪽으로 갈 수 있으면 무조건 왼쪽으로
            if(tc < c || (c - 1 >= 0 && Math.abs(r - tr) + Math.abs((c - 1) - tc) <= k)) {
                c--;
                answer.append("l");
                continue;
            }
            // 오른쪽으로 갈 수 있으면 무조건 오른쪽으로
            if(c > tc || (c + 1 < C && Math.abs(r - tr) + Math.abs((c + 1) - tc) <= k)) {
                c++;
                answer.append("r");
                continue;
            }
            // d, l, r 다 안되면 어쩔 수 없이 u
            r--;
            if(r < 0) {
                break;
            }
            answer.append("u");
        }
        
        return r == tr && c == tc ? answer.toString() : "impossible";
    }
}
